### PR TITLE
Create base pebble storage

### LINKED ISSUE #4606: Move register DB in Archive node to Access Node
## Problem
The first part of the unified edge node will be to have register data onboard for a future FVM.

## Solution
1. Move the pebble storage module https://github.com/onflow/flow-archive/tree/master/service/storage2 to the AN code base
2. Have Execution data sync index the entries to is as part of the ingestion engine.
3. Create new Exec data endpoints that serve this register data

### PR DESCRIPTION
**Pull request by koko1123**
This PR creates a base pebble implementation with Unit tests and adds it to the storage module.
closes #4606 
Depends on https://github.com/onflow/flow-go/pull/4656

### CODE REVIEW

File: storage/pebble/registers/payload/lookup.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/onflow/flow-go/storage/pebble/registers/config"
+)
+
+// lookupKey is the encoded format of the storage key for looking up register value
+type lookupKey struct {
+	encoded []byte
+}
+
+// newLookupKey takes a height and registerID, returns the key for storing the register value in storage
+func newLookupKey(height uint64, reg flow.RegisterID) *lookupKey {
```
Comment (sideninja) [inline storage/pebble/registers/payload/lookup.go:18]: I'm pretty sure we will need to add here a prefix for `registers`. So we can have multiple entities in the future in the same db.

File: storage/pebble/registers.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	}
+
+	return &Registers{
+		Payloads: payloadStor,
+	}, nil
+}
+
+func (l *Registers) Close() (err error) {
+	multierr.AppendInto(&err, l.Payloads.Close())
+	return
```
Comment (zhangchiqing) [inline storage/pebble/registers.go:39]:
```suggestion
	return l.Payloads.Close()
```

Doesn't seem to have multiple errors

File: storage/pebble/registers/payload/lookup.go (hunk 1 above)
Comment (zhangchiqing) [reply storage/pebble/registers/payload/lookup.go:18]: +1

File: storage/pebble/registers.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/cockroachdb/pebble"
+	"go.uber.org/multierr"
+
+	"github.com/onflow/flow-go/storage/pebble/registers/payload"
+)
+
+// library that implements pebble storage for registers
+type Registers struct {
+	Payloads *payload.Payloads
```
Comment (zhangchiqing) [inline storage/pebble/registers.go:15]: We should rename payloads into registers, because we are not storing payloads, but actually registers

File: storage/registers.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+// RegisterReader defines read-only operations on the register index.
+type RegisterReader interface {
+	// Get register by the register ID at a given block height.
+	//
+	// If the register at the given height was not indexed, returns the highest
+	// height the register was indexed at.
+	// An error is returned if the register was not indexed at all. Expected errors:
+	// - storage.ErrNotFound if the register was not found in the db
```
Comment (zhangchiqing) [inline storage/registers.go:25]: We need to handle the additional two cases described in `storage.ErrOutOfRange`
```suggestion
	// - storage.ErrNotFound if the register was not found in the db
	// - storage.ErrOutOfRange if the queried height is below the root block height or above the last indexed height.
```

File: storage/pebble/registers/payload/lookup.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
+package payload
```
Comment (zhangchiqing) [inline storage/pebble/registers/payload/lookup.go:1]: rename into `register`

File: storage/pebble/registers/payload/payload.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import (
+	"fmt"
+
+	"github.com/cockroachdb/pebble"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage/pebble/registers/config"
+)
+
+type Payloads struct {
```
Comment (zhangchiqing) [inline storage/pebble/registers/payload/payload.go:12]: rename into `Registers`

File: storage/pebble/registers.go (hunk 3 above)
Comment (zhangchiqing) [reply storage/pebble/registers.go:15]: This module is such a thin wrap, I'm not sure if we actually need it? 

I think we can move the actual DB operation to this module.

**Review by zhangchiqing — COMMENTED** [review]
I think the register storage should keep track of the latest height as well. as the root height, so that if the GetRegister query is beyond the range [root_height, latest_height] it should return out of range error. This can be done in a separate PR tho.

File: storage/pebble/registers/payload/lookup.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	// Encode the height getting it to 1s compliment (all bits flipped) and big-endian byte order.
+	// (Prefix iteration in pebble does not support reverse iteration.)
+	onesCompliment := ^height
+	key.encoded = binary.BigEndian.AppendUint64(key.encoded, onesCompliment)
+
+	return &key
+}
+
+// lookupKeyToRegisterID takes a lookup key and decode it into height and RegisterID
+func lookupKeyToRegisterID(lookupKey []byte) (uint64, flow.RegisterID, error) {
```
Comment (sideninja) [inline storage/pebble/registers/payload/lookup.go:42]: why is this not a method on the lookupKey type? a lot of feature-envy

File: storage/registers.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// Todo: Update interface after https://github.com/onflow/flow-go/pull/4656/ gets merged
+
+// Registers defines methods for the register index.
+type Registers interface {
+	RegisterReader
+	RegisterWriter
+	io.Closer
```
Comment (zhangchiqing) [inline storage/registers.go:15]: I don't think the Registers module need to handle the Close method. 

Registers might not be the only data stored in pebble, if let's say, chunk data pack is also stored in pebble, then ChunkDataPacks would have to provide the same Close method?

The Close method is a database level method, so we can let the database creator hold the pebble db instance and call it.

For instance, the FlowNodeBuilder is responsible for creating the badger DB instance, as well as closing it during shut down.
https://github.com/onflow/flow-go/blob/master/cmd/scaffold.go#L890

Comment (sideninja) [reply storage/registers.go:15]: I believe this is outdated, I already removed the close for these reasons, but @koko1123 probably needs to re-merge the branch.

File: storage/pebble/registers.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+func StoragePath(dir string) string {
+	return path.Join(dir, "payload.db")
+}
+
+func NewRegisters(dir string, blockCacheSize int64) (*Registers, error) {
+	cache := pebble.NewCache(blockCacheSize)
+	defer cache.Unref()
```
Comment (peterargue) [inline storage/pebble/registers.go:24]: is this safe? The [docs](https://github.com/cockroachdb/pebble/blob/c872ea071a9f1eeb3b3fed09fdb1bb028ff5f6fb/internal/cache/clockpro.go#L783-L784) say
> The cache only remains valid as long a reference is maintained to it.

so it seems like the cache would be immediately invalidated when returning from this method.

Comment (koko1123) [reply storage/pebble/registers.go:24]: This is an optimization from @SaveTheRbtz I think

Comment (rbtz-openai) [reply storage/pebble/registers.go:24]: [It should be fine](https://pkg.go.dev/github.com/cockroachdb/pebble#NewCache):

>  Each DB it is associated with adds a reference, so the creator of the cache should usually release their reference after the DB is created.

File: storage/pebble/registers/payload/lookup.go (hunk 7 above)
Comment (koko1123) [reply storage/pebble/registers/payload/lookup.go:42]: I could move it there, it's a private implementation on RN because of its single usecase

File: storage/pebble/registers.go (hunk 9 above)
Comment (peterargue) [reply storage/pebble/registers.go:24]: ah, thanks!

Comment (koko1123) [reply storage/pebble/registers.go:24]: wow I didn't think summoning him would work!

File: storage/height.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+type HeightIndex interface {
+	// LatestHeight returns the latest indexed height.
+	LatestHeight() (uint64, error)
+	// FirstHeight at which we started to index. Returns the first indexed height found in the store.
+	FirstHeight() (uint64, error)
+	// SetLatestHeight updates the latest height.
+	// The provided height should either be one higher than the current height or the same to ensure idempotency.
+	// If the height is not within those bounds it will panic!
+	// An error might get returned if there are problems with persisting the height.
+	SetLatestHeight(height uint64) error
```
Comment (zhangchiqing) [inline storage/height.go:14]: Why do we need this method? I think we agreed to set latest height within the Store method, no?

File: storage/pebble/registers/config/config.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
+package config
```
Comment (zhangchiqing) [inline storage/pebble/registers/config/config.go:1]: This is the config for pebble, not for registers, we could move it to storage/pebble/config/config.go

File: storage/height.go (hunk 10 above)
Comment (koko1123) [reply storage/height.go:14]: sorry did not update from Gregor's change

File: storage/pebble/lookup.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+	// The "<owner>/<key>" part is the register key, which is used as a prefix to filter and iterate
+	// through updated values at different heights, and find the most recent updated value at or below
+	// a certain height.
+	key.encoded = append(key.encoded, []byte(reg.Owner)...)
+	key.encoded = append(key.encoded, '/')
+	key.encoded = append(key.encoded, []byte(reg.Key)...)
+	key.encoded = append(key.encoded, '/')
+
+	// Encode the height getting it to 1s compliment (all bits flipped) and big-endian byte order.
+	// (Prefix iteration in pebble does not support reverse iteration.)
```
Comment (peterargue) [inline storage/pebble/lookup.go:35]: It took me a little bit to decode why we needed to use 1s complement here, so I write up some notes. Does this capture the rational?
```suggestion
	// Encode the height getting it to 1s compliment (all bits flipped) and big-endian byte order.
	//
	// Register are a sparse dataset stored with a single entry per update. To find the value at a particular
	// height, we need to do a scan across the entries to find the highest height that is less than or equal
	// to the target height.
	//
	// Pebble does not support reverse iteration, so we use the height's one's complement to effectively
	// reverse sort on the height. This allows us to use a bitwise forward scan for the next most recent
	// entry.
```

File: storage/pebble/registers_test.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+	expectedValue1 := []byte("value1")
+	entries := flow.RegisterEntries{
+		{Key: key1, Value: expectedValue1},
+	}
+
+	minHeight := uint64(2)
+	err = s.Store(minHeight, entries)
+	require.NoError(t, err)
+
+	value1, err := s.Get(minHeight, key1)
```
Comment (peterargue) [inline storage/pebble/registers_test.go:43]:
```suggestion
	// lookup with exact height returns the correct value
	value1, err := s.Get(minHeight, key1)
```

File: storage/pebble/registers_test.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+	require.NoError(t, err)
+	require.Equal(t, expectedValue1, value1)
+
+	value1, err = s.Get(minHeight+1, key1)
```
Comment (peterargue) [inline storage/pebble/registers_test.go:47]:
```suggestion
	// lookup with a higher height returns the correct value
	value1, err = s.Get(minHeight+1, key1)
```

File: storage/pebble/registers_test.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+	require.NoError(t, err)
+	require.Equal(t, expectedValue1, value1)
+
+	value1, err = s.Get(minHeight-1, key1)
```
Comment (peterargue) [inline storage/pebble/registers_test.go:51]:
```suggestion
	// lookup with a lower height returns no results
	value1, err = s.Get(minHeight-1, key1)
```

File: storage/pebble/registers.go [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+
+	err := batch.Commit(pebble.Sync)
+	if err != nil {
+		return fmt.Errorf("failed to commit batch: %w", err)
+	}
+
+	return nil
+}
+
+// TODO: Finish implementation after deciding prefixes and/or badger use
```
Comment (peterargue) [inline storage/pebble/registers.go:83]: what's left to do here?

Comment (koko1123) [reply storage/pebble/registers.go:83]: I'll combine this with the checkpoint implementation, since it involves s check for checkpoint ingestion completing, I'll  define a couple of constants for the keys and that should serve our purpose

File: storage/pebble/registers.go [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+	return valueCopy, nil
+}
+
+// Store sets the given entries in a batch.
+func (s *Registers) Store(
+	height uint64,
+	entries flow.RegisterEntries,
+) error {
+	batch := s.db.NewBatch()
```
Comment (zhangchiqing) [inline storage/pebble/registers.go:63]: is the height check going to be added in another PR? 

If so, please add a TODO here

File: storage/pebble/registers.go [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+//
+// For example, if there are 2 values stored for register A at height 6 and 11, then
+// GetPayload(13, A) would return the value at height 11.
+//
+// If no payload is found, an empty byte slice is returned.
+func (s *Registers) Get(
+	height uint64,
+	reg flow.RegisterID,
+) ([]byte, error) {
+	iter, err := s.db.NewIter(&pebble.IterOptions{
```
Comment (zhangchiqing) [inline storage/pebble/registers.go:33]: is the height check going to be added in another PR?

If so, please add a TODO here

**Review by zhangchiqing — APPROVED** [review]

**Review by peterargue — APPROVED** [review]

File: storage/pebble/registers.go [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	return nil
+}
+
+// TODO: Finish implementation after deciding prefixes and/or badger use
+
+// LatestHeight Gets the latest height of complete registers available
+func (s *Registers) LatestHeight() (uint64, error) {
+	return 0, nil
```
Comment (sideninja) [inline storage/pebble/registers.go:89]: Can we instead panic with not implemented since this is not yet implemented.
